doc: add information about s2n-tls software architecture (#4868)
aws/s2n-tls: docs/usage-guide/topics/ch04-connection.md

# TLS Connections

> Users will need to create a `s2n_connection` struct to store all of the state necessary for a TLS connection. One `s2n_connection` must be created for each TCP stream. Call `s2n_connection_new()` to create a new server or client connection. Call `s2n_connection_free()` to free the memory allocated for this struct when no longer needed.

Reviewer: We've sort of lost the original point of the PR now, which was to provide an overview of "s2n-tls software architecture". In my mind, that looks something like: "In general, s2n-tls works by operating on s2n_connections. A user should create a connection, associate a s2n_config with that connection, and then calling s2n_negotiate on the connection until the TLS handshake completes. Then users can call s2n_send/s2n_recv on the connection object to send and receive application data."

It's fine to provide that sort of general overview in the beginning of the connections.md file.

> Users will need to create a `s2n_connection` struct to store all of the state necessary for a TLS connection. Call `s2n_connection_new()` to create a new server or client connection. Call `s2n_connection_free()` to free the memory allocated for this struct when no longer needed. In a TLS server, there will be one `s2n_connection` for each TCP stream.

Reviewer 2: nit: s2n_connections are created for clients and servers, so specifying only servers might be confusing. I also think this sentence might fit better earlier:
```suggestion
Users will need to create a `s2n_connection` struct to store all of the state necessary for a TLS connection. One `s2n_connection` must be created for each TCP stream. Call `s2n_connection_new()` to create a new server or client connection. Call `s2n_connection_free()` to free the memory allocated for this struct when no longer needed.
```

# TLS Connections

Users interact with s2n-tls primarily through the `s2n_connection` and `s2n_config` structures. Users should build a `s2n_config` object and `s2n_connection` objects, associate the config with those connections, and then calling `s2n_negotiate()` on those connections until the TLS handshakes are completed. Users should then call `s2n_send`/`s2n_recv` on those connections to send and receive application data.

> 

Author: > I'm not really sure what the goal of this paragraph is or that its inclusion makes the documentation better.

The initial intent of this paragraph is to provide a general overview of S2N software architecture. I think such a high level overview does make the documentation better, since it provides a general idea of how S2N-TLS usually works. It also connects the following chapters together. This is also suggested by [this comment](https://github.com/aws/s2n-tls/pull/4868#discussion_r1831531265).

> technically you don't have to create a config to use a connection.

I would change the wording as `In general, users build a s2n_config ...` to indicate how S2N is generally configured. That change implies there can be exceptions.

> I'm also not sure how useful the very brief mention of negotiate and send/receive is-- if we're going to mention them, we should probably at least link to the usage guide section that goes into more details.

I think the process of negotiation and send/receive is a part of how users would typically interact with S2N-TLS. I would keep this part. I agree that we need to link chapter 7 - sending and receiving to this part.

> 

Author: > You may need to think about reorganizing things. Introduce what a connection is, then direct customers to the config section for configuration options and the IO section for how to use IO.

I want to set up a subsection under `# TLS Connections` called `## Software Architecture`, and move the paragraph into that subsection. Hence, the `s2n_connection` introduction will go before the general overview of software architecture.

Moreover, `## Connection Memory` subsection talks about freeing memory after handshake.
https://github.com/aws/s2n-tls/blob/01e90bb34829aa2d63c2c74534e33474085c3422/docs/usage-guide/topics/ch04-connection.md?plain=1#L9
Hence, introducing the software architecture after `# TLS Connections` and right before `## Connection Memory` should helps the doc to explain when the sequence of events usually take place.

> 

Reviewer 3: I'm not convinced that "Software Architecture" really makes sense as its own section. There's not much to talk about other than connections with software architecture :P

I'd go for:
1. what is a connection
2. how is a connection used, high level with links to details
3. how do connections interact with configs, high level with links to details

> 

Author: I wrote a general introduction about TLS connections and configs, which includes answers to those questions. We can revise it based on the new commit.

> In general, s2n-tls works by operating on `s2n_connection` structs. A user should first create a connection by calling `s2n_connection_new()`. Then a [TLS handshake can be performed](./ch07-io.md#performing-the-tls-handshake) on the connection. Finally, the connection can be used to [send and receive encrypted data](./ch07-io.md#application-data). An `s2n_config` struct can be associated with a connection to provide [configuration outside of the default](./ch05-config.md).

Reviewer 3: I'm not sure about the phrase "configuration outside of the default". Minimally it should probably be "defaults",  or just "additional configuration"? Or "configure additional options"?
```suggestion
In general, s2n-tls works by operating on `s2n_connection` structs. A user should first create a connection by calling `s2n_connection_new()`. Then a [TLS handshake can be performed](./ch07-io.md#performing-the-tls-handshake) on the connection. Finally, the connection can be used to [send and receive encrypted data](./ch07-io.md#application-data). An `s2n_config` struct can be associated with a connection to [configure additional options](./ch05-config.md).
```

> A TLS connection is a secure and encrypted channel established between two communicating peers. A TLS connection is used to send and receive data, perform TLS handshake negotiations, send alert messages, and etc. Read [TLS Connections](./ch04-connection.md) for information about connections, and [Sending and Receiving](./ch07-io.md) for io interactions. TLS connections are configured by TLS configs, which contains a collection of TLS settings. TLS configs apply rules on TLS connections by defining connection specs, such as supported certificate authorities. Read [Configuring the Connection](./ch05-config.md) for information about TLS connection and config interactions.

Reviewer 3: You need to give introductions for s2n-tls connections, but a user of a TLS library can probably be assumed to know what a TLS connection is. It also looks like you're using "TLS connection" and "s2n-tls connection" interchangeably tho. And "TLS configs" aren't a standard part of TLS, those are an s2n-tls implementation detail.

And I also still don't think this paragraph will really be useful to a new user trying to understand the overall flow of the library. You give examples of possible things you can do with connections, but it's really not clear enough to be helpful. What do I need to do with a connection to use TLS? You also include a link to this section, which is odd.

You also go too deep into "TLS configs"-- configs have their own section, they should at most be referenced here.
